### PR TITLE
[Doc] Add documentation for sparse computation

### PR DESCRIPTION
**Pull request by yuanming-hu**

### CODE REVIEW

File: docs/lang/articles/advanced/sparse.md [hunk 1]
```diff
@@ -4,7 +4,69 @@ sidebar_position: 3
...
 # Sparse computation
 
-The LLVM backends (CPU/CUDA) and the Metal backend offer the full functionality of spatially sparse computation in Taichi.
+## Motivation
+
+High-resolution 2D/3D grids are often needed in large-scale spatial computation, especially physical simulation.
+However, they tend to take a huge amount of memory space and computation.
+
+While a programmer may allocate large huge dense grids to store spatial data (especially physical quantities such as a density field),
+oftentimes he or she only cares about a small fraction of this dense grid, the rest being ambient space (air).
```
Comment (victoriacity) [inline docs/lang/articles/advanced/sparse.md:13]: It has become increasing popular in technical documentation to use the singular *they* as the general pronoun (e.g., [as in Google docs style guide](https://developers.google.com/style/pronouns)). I wonder if we shall also adopt this to make our wording fully inclusive to gender identities. [The singular *they* is used exactly the same as the regular plural *they* where third-person singular is not applied](https://en.wikipedia.org/wiki/Singular_they).

Comment (rexwangcc) [reply docs/lang/articles/advanced/sparse.md:13]: I think that is a great point and we should defo do it, but requires a huge effort to change all of the pronouns across our documentation, so maybe we could do it in a standalone PR? And while we are still on the topic of inclusion, our default branch is still not updated to `main` 😅

Comment (victoriacity) [reply docs/lang/articles/advanced/sparse.md:13]: 😂 Fortunately we haven't had any pronouns in the current documentation, if I've used our docs search correctly. Then the effort would be simply on writing the updated documentations.

File: docs/lang/articles/advanced/sparse.md [hunk 2]
```diff
@@ -4,7 +4,69 @@ sidebar_position: 3
 
 # Sparse computation
 
-The LLVM backends (CPU/CUDA) and the Metal backend offer the full functionality of spatially sparse computation in Taichi.
+## Motivation
+
+High-resolution 2D/3D grids are often needed in large-scale spatial computation, especially physical simulation.
+However, they tend to take a huge amount of memory space and computation.
+
+While a programmer may allocate large huge dense grids to store spatial data (especially physical quantities such as a density field),
```
Comment (victoriacity) [inline docs/lang/articles/advanced/sparse.md:12]:
```suggestion
While a programmer may allocate large dense grids to store spatial data (especially physical quantities such as a density field),
```

File: docs/lang/articles/advanced/sparse.md [hunk 3]
```diff
@@ -4,7 +4,69 @@ sidebar_position: 3
...
+oftentimes he or she only cares about a small fraction of this dense grid, the rest being ambient space (air).
+
+In short, the regions of interest may only occupy a small fraction of the bounding volume.
+If we can leverage such "spatial sparsity" and focus computation on the regions we care about,
+we significantly save storage and computing power.
```
Comment (victoriacity) [inline docs/lang/articles/advanced/sparse.md:17]:
```suggestion
we will significantly save storage and computing power.
```
Not necessary but I personally feel this is a more natural flow of language.

File: docs/lang/articles/advanced/sparse.md [hunk 4]
```diff
@@ -4,7 +4,69 @@ sidebar_position: 3
...
+
+## Sparse data structures in Taichi
+
+### Data structure hierarchy
+
+To concentrate computation on sparse regions of interest, multilevel sparse voxel data structures are studied extensively. 
+
+## Activation on write
+
+When writing to an inactive cell on a sparse data structre, Taichi automatically populate the data structure.
```
Comment (victoriacity) [inline docs/lang/articles/advanced/sparse.md:51]:
```suggestion
When writing to an inactive cell on a sparse data structre, Taichi automatically populates the data structure.
```

File: docs/lang/articles/advanced/sparse.md [hunk 5]
```diff
@@ -4,7 +4,69 @@ sidebar_position: 3
...
+
+![image](https://raw.githubusercontent.com/taichi-dev/public_files/master/taichi/sparse_grids.gif)
+
+## Sparse struct-fors
+
+Efficiently looping over sparse grid cells that are irregular can be a challenge. Fortunately, in Taichi, *struct-for's*
+natively support sparse data structures, and only loops over voxels that are currently active.
```
Comment (victoriacity) [inline docs/lang/articles/advanced/sparse.md:58]:
```suggestion
natively support sparse data structures and only loops over voxels that are currently active.
```
If we would like to have really polished grammar🤣: the subject of the part after "and" is also "*struct-for's*" thus [the grammatically correct form is without the comma](https://www.grammarly.com/blog/comma-before-and/).

File: docs/lang/articles/advanced/sparse.md (hunk 1 above)
Comment (yuanming-hu) [reply docs/lang/articles/advanced/sparse.md:13]: Sounds great and I will adopt singular *they* here. Today I learn something!

**Comment by yuanming-hu** [comment]
Progress: 60%

(Not ready for review yet... need more text, figures, and proofreading.)

**Comment by yuanming-hu** [comment]
Hopefully this initial version of sparse computation doc is not too messy: https://deploy-preview-2983--jovial-fermat-aa59dc.netlify.app/docs/lang/articles/advanced/sparse

File: docs/lang/articles/advanced/sparse.md [hunk 6]
```diff
@@ -4,12 +4,310 @@ sidebar_position: 3
...
+![image](https://raw.githubusercontent.com/taichi-dev/public_files/master/taichi/doc/sparse_grids_3d.jpg)
+Figure: A 3D fluid simulation that uses both particles and grids. Left to right: particles, 1x1x1 voxels, 4x4x4 blocks, 16x16x16 blocks.
+
+#### Blocked leaf cells and bitmasks
+
+While a null pointer can effectively represent an empty sub-tree, at the leaf level using 64 bits to represent the activity
+of a single voxel can consume too much space.
+For example, if each voxel contains a single `f32` value (4 bytes),
+the 64-bit pointer pointing to the value would take 8 bytes.
+The fact that storage costs of pointers are higher than the space to store the value themselves,
```
Comment (victoriacity) [inline docs/lang/articles/advanced/sparse.md:104]:
```suggestion
The fact that storage costs of pointers are higher than the space to store the value themselves
```

File: docs/lang/articles/advanced/sparse.md [hunk 7]
```diff
@@ -4,12 +4,310 @@ sidebar_position: 3
...
+goes against our goal to use sparse data structures to save space.
+
+To amortize the storage cost of pointers, programmers usually organize voxels in a *blocked* manner,
```
Comment (victoriacity) [inline docs/lang/articles/advanced/sparse.md:107]:
```suggestion
To amortize the storage cost of pointers, programmers usually organize voxels in a *blocked* manner
```

File: docs/lang/articles/advanced/sparse.md [hunk 8]
```diff
@@ -4,12 +4,310 @@ sidebar_position: 3
...
+- does **not** trigger an deactivation of its parent container, even if all the children of the parent container are deactivated.
+:::
+
+:::note
+When deactivation happens, the Taichi runtime automatically recycles and zero-fills memory of the deactivated containers.
+:::
+
+:::note
+While it is possible to directly use `[i // 2, j // 2]` to compute the `block` index given `pixel` index,
+doing so couples computation code with the internal configuration of data structures (in this case, the size of `block` containers).o
```
Comment (victoriacity) [inline docs/lang/articles/advanced/sparse.md:300]:
```suggestion
doing so couples computation code with the internal configuration of data structures (in this case, the size of `block` containers).
```

**Review by victoriacity — APPROVED** [review]
LGTM! Just suggested a few trivial edits.

**Pull request merged**